add --watch option to `tele list`

njayp opened the pull request:
Signed-off-by: njayp <nickjaypowell@gmail.com>

## Description

A `WatchWorkloads` grpc already exists, i add a flag in `tele list` that uses it.

Reproducer: 
Run `telepresence list -w`
Hit `ctrl c`
Run a telepresence command. All commands will hang.

njayp committed:
add --watch option to `tele list`

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
changelog

Signed-off-by: njayp <nickjaypowell@gmail.com>

raphaelreyna reviewed (commented):
This might cause some issues with the global `--output` flag since it buffers prints to stdout when outputing json. 
Perhaps check if that global flag is set to `json` and if so return an error or just set the local `--json` flag?

njayp committed:
ensure watchers, no grpc cleanup

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
lower cog complex

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp commented:
so i understand better, what issues would it cause?

njayp committed:
make linter happy

Signed-off-by: njayp <nickjaypowell@gmail.com>

raphaelreyna commented:
> so i understand better, what issues would it cause?

When the `--output=json` flag is set, the `io.Writer` returned by `cmd.OutOrStdout()` is not backed by `os.Stdout` but by a `strings.Builder` which is then used to form the json response that is flushed to `os.Stdout` _after_ the command exits.
Basically, it prevents any sort of "streaming" output.

thallgren commented on pkg/client/cli/cmd_list.go:
```diff
@@ -100,8 +118,17 @@ func (s *listInfo) list(cmd *cobra.Command, _ []string) error {
 		}
 	}
 
+	if len(workloads) == 0 {
```
Nit, but shouldn't this test be the very first thing that happens in the function?

thallgren commented on pkg/client/cli/cmd_list.go:
```diff
@@ -69,22 +71,38 @@ func (s *listInfo) list(cmd *cobra.Command, _ []string) error {
+		if err != nil {
+			return err
+		}
+		/*
+			go func() {
+				<-ctx.Done()
+				stream.CloseSend()
+			}()
+		*/
```
Please remove this comment, or explain why it's there.

thallgren reviewed (changes requested):
Looks good overall. Perhaps add an integration test to verify the functionality?

njayp committed:
address feedback and change --json to --snapshot

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp commented:
> I can see how the term "snapshot" makes sense for a watcher that delivers snapshots, but it's a bit far fetched as an option for a list command.
> 
> * What's the difference when you list with or without a `--snapshot`?
> * When you use `--snapshot` you get JSON output.
> * Oh? I though I'd get that with the global `--json` flag?
> * Yeah, you will, but that's different, because <fill in explanation here>.

My thinking was `--output json` gets you what you usually get, but in a json obj. `--snapshot` gets you the full snapshot object. Both gets you the snapshot object in your json object. I can explain this in the changelog as a change

thallgren commented:
@njayp "Both gets you the snapshot object in your json object" sounds like the `--snapshot` option is redundant, which I think it somehow isn't?

njayp commented:
> @njayp "Both gets you the snapshot object in your json object" sounds like the `--snapshot` option is redundant, which I think it somehow isn't?

its not redundant because errors. without the `--output json` flag set, errors are not in json format

thallgren commented:
Ok? And that's an distinction that is important enough to introduce this somewhat confusing flag?

njayp commented:
> Ok? And that's an distinction that is important enough to introduce this somewhat confusing flag?

i need json errors for docker. what we could do is detect if `--output json` is set, and if it is, print the snapshot object in the json object. This would remove the ability to output the human-readable list in a json object, but maybe that isn't a common use-case. Is this what you would prefer?

thallgren commented:
Ok, so why don't we change the `--json` flag to give us JSON errors then? @raphaelreyna would that be hard to do?

thallgren commented:
What I don't want is different ways of telling a command to produce JSON output, motivated needs that our users rarely know anything about.

njayp commented:
> Ok, so why don't we change the `--json` flag to give us JSON errors then? @raphaelreyna would that be hard to do?

I am against having multiple json flags floating around, imo having `--json` and `--output json` have different behavior is more confusing

thallgren commented:
I totally agree. I think we should have one single way of asking for JSON output. The only viable use-case for it is when the CLI is called from another process, and when that happens, surely both output and errors must be returned as JSON.

raphaelreyna committed:
structured output streaming (#2602)

* Make traffic-agents configuration follow suit when things change

A service that is removed or added will often affect the set of possible
routes that can be intercepted. This commit adds a watcher for services
and logic to discover what existing entries that may be affected by
the service addition/modification/deletion.

In addition, the mutating webhook now also listens for pod deletions, so
that the corresponding configmap entry can be removed when the owner
workload cease to exist.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Make intercept state know about all possible intercepts

An intercept state must know about all possible intercepts (i.e. the
intercepts using the same port) in order to be matchable against an
intercept specification.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Update changelog.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Fix dlog.Warn -> dlog.Warnf

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Fix problem with watcher not terminating correctly when cancelled

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Fix problem uninstall --agent nnn or --all-agents

A cluster-wide watcher would be started when no namespace was given.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* chart

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Fix problem with config being recreated by webhook delete event.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Ensure proper cleanup in integration_tests

The other-NNN namespace and the Helm install in the custom-NNN namespace
wasn't cleaned up properly.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Make daemon session terminate properly when traffic-manager dies

A Helm uninstall of the traffic-manager left the root daemon in a bad
state, constantly trying to reconnect to obtain cluster info. This
commit ensures that the session dies.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Handle relative mount points

The `telepresence intercept --mount <mountPoint>` now handles relative
paths on non-windows platforms. Windows will still require the path to
be a drive letter followed by a colon.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

* Refactor intercept cleanup logic to be done on removeIntercept.

It really made no sense to do it in a separate GC goroutine when the
user's session may already be terminated and it may be no longer be
possible to use the user's credentials to contact systema for cleanup.

Signed-off-by: Jose Cortes <josecortes@datawire.io>

* Don't store intercept API keys in a separate map.

Doesn't really make sense to since intercept termination will happen
with the client's credentials.

Signed-off-by: Jose Cortes <josecortes@datawire.io>

* Return codes.NotFound when no intercept is found

Co-authored-by: Thomas Hallgren <thomas@datawire.io>
Signed-off-by: Jose Cortes <josecortes@datawire.io>

* Address PR feedback

Signed-off-by: Jose Cortes <josecortes@datawire.io>

* add structured streaming to pkg output

Signed-off-by: Raphael Reyna <raphaelreyna@protonmail.com>

* Add StructuredStreamer unit test

Signed-off-by: Raphael Reyna <raphaelreyna@protonmail.com>

* Add more coverage around StrcutredStreamer

Signed-off-by: Raphael Reyna <raphaelreyna@protonmail.com>

Co-authored-by: Thomas Hallgren <thomas@datawire.io>
Co-authored-by: Jose Cortes <josecortes@datawire.io>

njayp committed:
updated to use streamer

Signed-off-by: njayp <nickjaypowell@gmail.com>

thallgren commented:
Something weird happened in PR #2602. It contains a whole slew of commits that doesn't belong to it and that affects the change-set of this PR (surely it doesn't span 32 files?). Some rebasing is needed in order to see what's included here.

njayp committed:
test

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
see stdout

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
use channel to pretty-print stdout

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
close when ctx canceled

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
labels

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
hard context

Signed-off-by: njayp <nickjaypowell@gmail.com>

njayp committed:
allow error when context killed

Signed-off-by: njayp <nickjaypowell@gmail.com>
